docs/testing-teal.md

```diff
@@ -0,0 +1,80 @@
+# Testing TEAL
+
+[TEAL](https://developer.algorand.org/docs/reference/teal/specification/) is a bytecode based stack language that executes inside Algorand transactions to check the parameters of the transaction and approve the transaction as if by a signature. `@algorand-builder/algorand-js` provides a simple workflow to execute TEAL code with high speed and throughput (as the transactions are not processed on the blockchain but in a test environment).
+
+**NOTE:** TEAL cannot modify or create transactions, only reject or approve them. Approval is signaled by finishing with the stack containing a single non-zero uint64 value.
+
+## How it works
+
+`@algorand-builder/algorand-js` (JavaScript Algorand runtime) executes transactions and processes TEAL in 3 parts :-
+
+- [Runtime](../packages/algorand-js/src/runtime): For a transaction or txn group, the state management is handled by the `Runtime`. User can use `Runtime` class to set up accounts, create applications, opt-in to app, update app..etc
+- [Parser](../packages/algorand-js/src/parser): Reads TEAL code and converts it to a list of opcodes which are executable by the interpreter. If any opcode/data in teal code is invalid, parser will throw an error.
+- [Interpreter](../packages/algorand-js/src/interpreter): Executes the list of opcodes returned by parser and updates stack after each execution. After the end of execution, if top of stack contains a non-zero uint64 value then the teal code is approved, and transaction can be executed.
+
+## Run tests
+In this section we will demonstrating executing transactions with stateless and stateful teal.
+
+### Stateless TEAL
+
+Let's try to execute a transaction where the user (say `john`) can withdraw funds from an `escrow` account based on a stateless smart contract logic. Teal code can be found [here](../packages/algorand-js/test/fixtures/escrow-account/assets/escrow.teal).
+- First let's set up the state: initialize accounts and set up runtime (snippet from mocha test is also provided below).
+```
+const escrow = new StoreAccountImpl(1000e6); // 1000 ALGO
+const john = new StoreAccountImpl(500, johnAccount); // 0.005 ALGO
+const runtime = new Runtime([escrow, john]); // setup runtime
+```
+![image](https://user-images.githubusercontent.com/33264364/104197847-81afe200-544b-11eb-8553-b2446ea9c763.png)
+
+- Execute transaction (using `runtime.executeTx()`) with valid txnParams.
+![image](https://user-images.githubusercontent.com/33264364/104199023-f33c6000-544c-11eb-975b-1f8f72508ee6.png)
```

Comment:
Let's use code sections. It's easier to edit it in the future.

Reply:
ok